### PR TITLE
Use tempTable class in creation of dedupe table.

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Replace another instance of creating a temp table with the agreed construct

Before
----------------------------------------
No prefixing, stdised handling

After
----------------------------------------
prefixing, stdised handling

Technical Details
----------------------------------------
The is covered by unit tests ContactTest::testDuplicateTest and DedupeFinderTest::testUnsupervisedDupes

There are some others in this file but I focussed on keeping the scope to a single table replacement.

A couple of thoughts
1) I passed the indexes in as 'columns' - works, is tested so we'd know if anything broke it but it feels a bit ug
2) I think it makes sense to add a descontruct on the class to drop the temp tables
3) the longer syntax for referring to the table is a bit wordy - but harmless I guess & reformatting would
help

Comments
----------------------------------------
@mfb I wound up in this code today so thought I'd pick up one instance to change if you want to review it

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
fix for test in #13866

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by mfb** [comment]
> I passed the indexes in as 'columns' - works, is tested so we'd know if anything broke it but it feels a bit ug

There's nothing wrong with passing in indexes as "columns", which is actually "table definition" not columns.

> I think it makes sense to add a deconstruct on the class to drop the temp tables

In general I think we can rely on MySQL to automatically drop temporary tables at the end of the session.  However, if a process is creating a series of large temporary tables, we would want to drop them to avoid disk space issues.

**Comment by eileenmcnaughton** [comment]
yeah I don't think these temp tables cause probs

File: CRM/Dedupe/BAO/RuleGroup.php [hunk 1]
```diff
@@ -233,7 +236,7 @@ public function fillTable() {
           if ($searchWithinDupes) {
             // get prepared to search within already found dupes if $searchWithinDupes flag is set
             $dao->query("DROP TEMPORARY TABLE IF EXISTS dedupe_copy");
-            $dao->query("CREATE TEMPORARY TABLE dedupe_copy SELECT * FROM dedupe WHERE weight >= {$weightSum}");
+            $dao->query("CREATE TEMPORARY TABLE dedupe_copy SELECT * FROM {$this->temporaryTables['dedupe']} WHERE weight >= {$weightSum}");
```
Comment (mfb) [inline CRM/Dedupe/BAO/RuleGroup.php]: This could use TempTable::createWithQuery().

File: CRM/Dedupe/BAO/RuleGroup.php [hunk 2]
```diff
@@ -244,7 +247,7 @@ public function fillTable() {
             if (preg_match('/dedupe_copy[\S\s]*(union)[\S\s]*dedupe_copy/i', $query, $matches, PREG_OFFSET_CAPTURE)) {
               // Make a second temp table:
               $dao->query("DROP TEMPORARY TABLE IF EXISTS dedupe_copy_2");
-              $dao->query("CREATE TEMPORARY TABLE dedupe_copy_2 SELECT * FROM dedupe WHERE weight >= {$weightSum}");
+              $dao->query("CREATE TEMPORARY TABLE dedupe_copy_2 SELECT * FROM {$this->temporaryTables['dedupe']} WHERE weight >= {$weightSum}");
```
Comment (mfb) [inline CRM/Dedupe/BAO/RuleGroup.php]: This could use TempTable::createWithQuery().

**Review by mfb — CHANGES_REQUESTED** [review]

File: CRM/Dedupe/BAO/RuleGroup.php (hunk 1 above)
Comment (eileenmcnaughton) [reply CRM/Dedupe/BAO/RuleGroup.php]: Yeah - I agree - I haven't included here because that will require a new chunk of digging into ensuring that test lines are passing through that code

**Comment by eileenmcnaughton** [comment]
@mfb shall I merge this? I think your concern was just that it doesn't do more (which wasn't the intent)

**Review by mfb — APPROVED** [review]
Looks good, we can follow up later

**Comment by eileenmcnaughton** [comment]
thanks - I'll try to grab time to test a report or two from your changes later today

**Comment by magnolia61** [comment]
This is the only place I can find that relates to my issue. When performing semi-large dedupes I get the a fatal error with the following in the CiviCRM log:

`    [debug_info] => CREATE TEMPORARY TABLE `civicrm_tmp_e_dedupe_c5bbafc3177fc756e2e3a4e2d13e4bf4` ENGINE=InnoDB DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci AS SELECT * FROM civicrm_tmp_e_dedupe_20319758a91b7e9d521a24196dac2a40 WHERE weight >= 20 [nativecode=1050 ** Table 'civicrm_tmp_e_dedupe_c5bbafc3177fc756e2e3a4e2d13e4bf4' already exists]`

It looks like the temp tables do not get dropped correctly. Or would you think this has to do with something totally different?
